<a href="https://colab.research.google.com/github/PosgradoMNA/actividades-de-aprendizaje-AngeHdz/blob/main/Semana3_Actividad_1_A01794256.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Semana 3 - Actividad 1

Nombre de Alumno: Angel De Jesus Hernandez Pascual

N. matrícula: A01794256

Profesor:	Jobish Vallikavungal Devassia 

Fecha: 10/03/2022

#Parte 1: Fundamentos de bases de datos

1.   Fundamentos de bases de datos y para ciencia de datos.  
    Una base de datos es una recopilacion organizada de informacion o datos estructurados, que nomarmelmete se almacena de forma electronica en un sistema informatico. Normalmente este provee un denominado Freamwork que provee la estrucura, almacenamiento, y la recuperacion de datos, 
    Existen dos tipos de base de datos:

*   Relacionales (PostgresSQL, MySql, DB2)
    En la basases de datos relaciones la informacion esta organizada en tablas con elementos que pueden estar relacionados a otras tablas, la forma de interactuar con estas es usan el lenguaje SQL
*   No relacionales (RabbitMQ, MongoDB, JanusGraphs)
    Mejor concidas como NoSQL, las bases de datos no relacionales almacenan la informacion por categorias en lugar de relaciones. La forma mas popular de almacenamiento es utilizando Kay-Value muy similar a como lo hacen los diccionarios de python.Con una llave unica que da acceso a un denominado documento el cual esta relacionada a la llave. Se puede agregar, eliminar y realizar updates, siguiendo la estructura del documento.

2.   Fundamentos de almacenes de datos (Data Warehouse) para ciencia de datos
   Un Data Warehouse, es un systema que esta diseñado para almacenar datos y el analizis de estos, utilizando busines intelligence. Este sistema centraliza multiples datos, como datos en tiempo real, datos historicos y es conciderado com ola unica fuente de la verdad.


#Parte 3: Preparación de los datos

Con base en los resultados de tu libreta de Google Colab de la Parte 2 responde detalladamente las siguientes preguntas: 



1.   ¿Qué datos considero mas importantes? 
      Desde mi perspectiva lo mas importante es el monto del credito otorgado (X1), La edad (X5), el historia de pagos (X6-X11), el repaiment status(X7)
      
      ¿Por qué? 
      Basicamente sabes cual es el monto del credito otorgado nos ayuda a comprender que tan buen cleinte es, y el nivel de pago que tiene comparandolo con el historia de pagos, saber si tiene un monto pequeño y esta pagando al corriente es un potencial cliente.Por el contrario un monto pequeño y falta de pago es un cliente al que se le negara otro credito.

2.   ¿Se eliminaron o reemplazaron datos nulos? Para los rows, con datos faltantes en el hisoria de pago, estado de cuenta, edad se eliminaron.
¿Qué se hizo y por qué? Por que considero estos datos importantes, son los datos que se necesitan para realizar el modelo, Los datos que solo tenian un faltante se decidio reemplazarlos por la moda de su columna, debido a que la cantidad que el dato moda es el mas usado ene cada columna.

3.   ¿Es necesario ordenar los datos para el análisis? No ¿Por qué? En este caso solo realze la busqueda de los datos NaN, y revise sus columnas para saber cuantos datos me faltaban.

4.   ¿Existen problemas de formato que deban solucionar antes del proceso de modelado? NO Por qué? El 100% de los datos ya vienen en formato Float, solo tenia espacion vacios.

5.   ¿Qué ajustes se realizaron en el proceso de limpieza de datos (agregar, integrar, eliminar, modificar registros (filas), cambiar atributos (columnas)?

Para los rows con mas de 10 faltantes se decidio eliminarlos o que no tuviesen el historial de pagos o saldos.
Para los rows con un solo faltante se reviso el la moda de la columna, la cual tendia a 0. asi que se utilizo el valor moda como reemplazo.
Ejemplo  la comuna X14 con los datos: 
ndf['X14'].value_counts()

0.0         2868

390.0        275

780.0         73

326.0         63

316.0         62

            ... 

12438.0        1

19580.0        1

45129.0        1

227807.0       1

49764.0        1

Name: X14, Length: 22020, dtype: int64

Donde no se puede utilizar el promedio o un cuartil ya no estaria balanceado, se decidio utilizar la moda ya que es el valor mas incidente.






#Parte 2: Selección y limpieza de los Datos en Python

In [208]:
link = "https://raw.githubusercontent.com/PosgradoMNA/Actividades_Aprendizaje-/main/default%20of%20credit%20card%20clients.csv"

In [209]:
import pandas as pd
import numpy as np
df = pd.read_csv(link)

In [210]:
df['X2'].value_counts()

2.0    18112
1.0    11887
Name: X2, dtype: int64

In [211]:
df.isna().sum()

ID      0
X1      0
X2      1
X3      2
X4      2
X5      5
X6      3
X7      5
X8      7
X9      9
X10    16
X11    14
X12    11
X13    11
X14    13
X15    15
X16    17
X17    10
X18     8
X19     9
X20     8
X21    11
X22    11
X23     5
Y       3
dtype: int64

Columna X2 se eliminara el ID 24366 debido a que contiene solo el ID y X1

In [212]:
df[df['X2'].isna()]

ID      X1  X2  X3  X4  X5  X6  X7  X8  X9  ...  X15  X16  X17  X18  \
24365  24366  130000 NaN NaN NaN NaN NaN NaN NaN NaN  ...  NaN  NaN  NaN  NaN   

       X19  X20  X21  X22    X23    Y  
24365  NaN  NaN  NaN  NaN  390.0  0.0  

[1 rows x 25 columns]

In [213]:
#Se crea una copia para trabajar
ndf = df.copy()

In [214]:
# Se eliminan el row debido a que contienen muchos datos vacios
ndf.dropna(subset=["X2"], inplace= True)

In [215]:
ndf[ndf['X2'].isna()]

Empty DataFrame
Columns: [ID, X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13, X14, X15, X16, X17, X18, X19, X20, X21, X22, X23, Y]
Index: []

[0 rows x 25 columns]

Columna X3 el ID 6228 contien mas de 10 datos vacios, tambien se eliminara

In [216]:
ndf[ndf['X3'].isna()]

ID     X1   X2  X3  X4  X5  X6  X7  X8  X9  ...  X15  X16      X17  \
6227  6228  30000  1.0 NaN NaN NaN NaN NaN NaN NaN  ...  NaN  NaN  21120.0   

         X18     X19     X20     X21     X22  X23    Y  
6227  1586.0  1365.0  1663.0  5106.0  1050.0  0.0  0.0  

[1 rows x 25 columns]

In [217]:
ndf.dropna(subset=["X3"], inplace= True)

In [218]:
#Se revisa cuantos datos vacios llevamos
ndf.isna().sum()

ID      0
X1      0
X2      0
X3      0
X4      0
X5      3
X6      1
X7      3
X8      5
X9      7
X10    14
X11    12
X12     9
X13     9
X14    11
X15    13
X16    15
X17     9
X18     7
X19     8
X20     7
X21    10
X22    10
X23     5
Y       3
dtype: int64

La columna X5 los ID 83,220 y 240 se reemplazaran por el promedio de la misma columna X5

In [219]:
ndf[ndf['X5'].isna()]

ID      X1   X2   X3   X4  X5   X6   X7   X8   X9  ...       X15  \
182  183  500000  2.0  1.0  1.0 NaN  0.0  0.0  0.0  0.0  ...  122967.0   
219  220  310000  2.0  1.0  2.0 NaN -1.0 -1.0 -1.0 -1.0  ...       0.0   
239  240  140000  2.0  2.0  3.0 NaN  0.0  0.0  0.0  0.0  ...   19068.0   

          X16      X17      X18      X19      X20      X21      X22      X23  \
182  108834.0  70064.0  70010.0  30357.0  30000.0  20000.0  52183.0  20000.0   
219       0.0      0.0   4542.0    126.0      0.0      0.0      0.0      0.0   
239   16409.0  16383.0   3000.0   2000.0   2198.0   1000.0   3000.0   2399.0   

       Y  
182  0.0  
219  0.0  
239  1.0  

[3 rows x 25 columns]

In [220]:
ndf.describe()

ID              X1            X2            X3            X4  \
count  29998.000000    29998.000000  29998.000000  29998.000000  29998.000000   
mean   15000.480232   167490.155344      1.603774      1.853057      1.551903   
std     8660.370150   129749.377992      0.489121      0.790320      0.521968   
min        1.000000    10000.000000      1.000000      0.000000      0.000000   
25%     7501.250000    50000.000000      1.000000      1.000000      1.000000   
50%    15000.500000   140000.000000      2.000000      2.000000      2.000000   
75%    22499.750000   240000.000000      2.000000      2.000000      2.000000   
max    30000.000000  1000000.000000      2.000000      6.000000      3.000000   

                 X5            X6            X7            X8            X9  \
count  29995.000000  29997.000000  29995.000000  29993.000000  29991.000000   
mean      35.484214     -0.016635     -0.133689     -0.166405     -0.220800   
std        9.218024      1.123829      1.197254      1.196048      1.169153   
min       21.000000     -2.000000     -2.000000     -2.000000     -2.000000   
25%       28.000000     -1.000000     -1.000000     -1.000000     -1.000000   
50%       34.000000      0.000000      0.000000      0.000000      0.000000   
75%       41.000000      0.000000      0.000000      0.000000      0.000000   
max       79.000000      8.000000      8.000000      8.000000      8.000000   

       ...            X15            X16            X17            X18  \
count  ...   29985.000000   29983.000000   29989.000000   29991.000000   
mean   ...   43275.652326   40324.493980   38881.728000    5663.081825   
std    ...   64345.500073   60809.984983   59562.217735   16564.424518   
min    ... -170000.000000  -81334.000000 -339603.000000       0.000000   
25%    ...    2329.000000    1763.500000    1256.000000    1000.000000   
50%    ...   19052.000000   18107.000000   17080.000000    2100.000000   
75%    ...   54560.000000   50213.000000   49209.000000    5006.000000   
max    ...  891586.000000  927171.000000  961664.000000  873552.000000   

                X19            X20            X21            X22  \
count  2.999000e+04   29991.000000   29988.000000   29988.000000   
mean   5.922641e+03    5225.742189    4827.243231    4800.422269   
std    2.304455e+04   17608.704177   15669.013149   15281.081511   
min    0.000000e+00       0.000000       0.000000       0.000000   
25%    8.352500e+02     390.000000     296.000000     251.000000   
50%    2.009000e+03    1800.000000    1500.000000    1500.000000   
75%    5.000000e+03    4506.000000    4014.000000    4033.750000   
max    1.684259e+06  896040.000000  621000.000000  426529.000000   

                 X23             Y  
count   29993.000000  29995.000000  
mean     5216.594805      0.221204  
std     17779.393778      0.415064  
min         0.000000      0.000000  
25%       118.000000      0.000000  
50%      1500.000000      0.000000  
75%      4000.000000      0.000000  
max    528666.000000      1.000000  

[8 rows x 25 columns]

In [221]:
ndf['X5'].fillna(ndf['X5'].mean(), inplace = True)

In [222]:
ndf[ndf['X5'].isna()]

Empty DataFrame
Columns: [ID, X1, X2, X3, X4, X5, X6, X7, X8, X9, X10, X11, X12, X13, X14, X15, X16, X17, X18, X19, X20, X21, X22, X23, Y]
Index: []

[0 rows x 25 columns]

In [223]:
#Se revisa cuantos datos vacios llevamos
ndf.isna().sum()

ID      0
X1      0
X2      0
X3      0
X4      0
X5      0
X6      1
X7      3
X8      5
X9      7
X10    14
X11    12
X12     9
X13     9
X14    11
X15    13
X16    15
X17     9
X18     7
X19     8
X20     7
X21    10
X22    10
X23     5
Y       3
dtype: int64

La columna x6 
el ID 2424 tambien contiene mas de 10 datos faltantes, sera elimada el row

In [224]:
ndf[ndf['X6'].isna()]

ID     X1   X2   X3   X4    X5  X6  X7  X8  X9  ...  X15  X16  X17  \
24123  24124  50000  1.0  1.0  2.0  29.0 NaN NaN NaN NaN  ...  NaN  NaN  NaN   

       X18  X19  X20  X21  X22    X23    Y  
24123  NaN  NaN  NaN  NaN  NaN  360.0  0.0  

[1 rows x 25 columns]

In [225]:
ndf.dropna(subset=["X6"], inplace= True)

In [226]:
#Se revisa cuantos datos vacios llevamos
ndf.isna().sum()

ID      0
X1      0
X2      0
X3      0
X4      0
X5      0
X6      0
X7      2
X8      4
X9      6
X10    13
X11    11
X12     8
X13     8
X14    10
X15    12
X16    14
X17     8
X18     6
X19     7
X20     6
X21     9
X22     9
X23     5
Y       3
dtype: int64

en la columna 

Columna X7 se eliminaran los ID 29833 y 29834 por que contiene mas de dies datos faltantes que son los datos de los pagos por mes los cuales sin importantes para el analisis

In [227]:
ndf[ndf['X7'].isna()]

ID     X1   X2   X3   X4    X5   X6  X7  X8  X9  ...  X15  X16  \
29832  29833  20000  1.0  2.0  2.0  30.0  0.0 NaN NaN NaN  ...  NaN  NaN   
29833  29834  20000  1.0  2.0  2.0  31.0 -1.0 NaN NaN NaN  ...  NaN  NaN   

           X17     X18     X19     X20    X21      X22     X23    Y  
29832  20090.0  1279.0  3000.0  6668.0  468.0    507.0  5582.0  0.0  
29833  19103.0   390.0   780.0     0.0  806.0  19103.0  1000.0  1.0  

[2 rows x 25 columns]

In [228]:
#Se eliminan los rows que en la columna X7 contienen datos perdidos
ndf.dropna(subset=["X7"], inplace= True)

In [229]:
#Se revisa cuantos datos vacios llevamos
ndf.isna().sum()

ID      0
X1      0
X2      0
X3      0
X4      0
X5      0
X6      0
X7      0
X8      2
X9      4
X10    11
X11     9
X12     6
X13     6
X14     8
X15    10
X16    12
X17     8
X18     6
X19     7
X20     6
X21     9
X22     9
X23     5
Y       3
dtype: int64

Columna X8 los ID 17990 y 18004 con valores NaN seran reemplazados por el valor como mayor repetibilidad (0.0)

In [230]:
ndf[ndf['X8'].isna()]

ID     X1   X2   X3   X4    X5   X6   X7  X8   X9  ...      X15  \
17989  17990  80000  1.0  2.0  1.0  46.0  0.0  0.0 NaN  0.0  ...  64237.0   
18003  18004  70000  1.0  2.0  2.0  33.0  1.0  2.0 NaN  7.0  ...  38922.0   

           X16      X17      X18     X19     X20     X21     X22     X23    Y  
17989  47890.0  48036.0   3542.0  3500.0  2500.0  2000.0  2500.0  2000.0  0.0  
18003  38318.0  38133.0  12000.0     0.0  2500.0     0.0  1000.0  1500.0  0.0  

[2 rows x 25 columns]

In [231]:
#Se contabilizan los datos en la columna X8
ndf['X8'].value_counts()

 0.0    15761
-1.0     5935
-2.0     4085
 2.0     3819
 3.0      240
 4.0       76
 7.0       27
 6.0       23
 5.0       21
 1.0        4
 8.0        2
Name: X8, dtype: int64

In [232]:
ndf['X8'].fillna(0.0, inplace = True)

In [233]:
ndf[ndf['ID'] == 17990]

ID     X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15  \
17989  17990  80000  1.0  2.0  1.0  46.0  0.0  0.0  0.0  0.0  ...  64237.0   

           X16      X17     X18     X19     X20     X21     X22     X23    Y  
17989  47890.0  48036.0  3542.0  3500.0  2500.0  2000.0  2500.0  2000.0  0.0  

[1 rows x 25 columns]

In [234]:
ndf[ndf['ID'] == 18004]

ID     X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15  \
18003  18004  70000  1.0  2.0  2.0  33.0  1.0  2.0  0.0  7.0  ...  38922.0   

           X16      X17      X18  X19     X20  X21     X22     X23    Y  
18003  38318.0  38133.0  12000.0  0.0  2500.0  0.0  1000.0  1500.0  0.0  

[1 rows x 25 columns]

In [235]:
#Se revisa cuantos datos vacios llevamos
ndf.isna().sum()

ID      0
X1      0
X2      0
X3      0
X4      0
X5      0
X6      0
X7      0
X8      0
X9      4
X10    11
X11     9
X12     6
X13     6
X14     8
X15    10
X16    12
X17     8
X18     6
X19     7
X20     6
X21     9
X22     9
X23     5
Y       3
dtype: int64

Columna X9 tambien se reemplazara los valores NaN a 0.0 debido a que es el valor que mas repite y es el unico valor que le falta a los Rows

In [236]:
#se bucan los valores vacios en la columna X9
ndf[ndf['X9'].isna()]

ID     X1   X2   X3   X4    X5   X6   X7   X8  X9  ...      X15  \
160      161  30000  1.0  1.0  2.0  41.0  2.0  2.0  2.0 NaN  ...  28168.0   
23990  23991  60000  1.0  2.0  2.0  26.0  0.0  0.0  0.0 NaN  ...  58938.0   
23991  23992  50000  1.0  3.0  2.0  25.0  0.0  0.0  0.0 NaN  ...  49485.0   
23992  23993  50000  1.0  2.0  2.0  24.0  0.0  0.0  0.0 NaN  ...  41286.0   

           X16      X17     X18     X19     X20     X21     X22     X23    Y  
160    27579.0  28321.0  3500.0     0.0  2200.0     NaN  1200.0  1250.0  0.0  
23990  30101.0  29735.0  2113.0  2500.0  3000.0  2500.0  1500.0  1100.0  0.0  
23991  49083.0  45187.0  2100.0  2200.0  3500.0  2000.0     0.0  5151.0  0.0  
23992  16764.0  16960.0  2150.0  2352.0  1239.0   717.0   585.0   589.0  0.0  

[4 rows x 25 columns]

In [237]:
#Se revisa que valores contiene la columna determinando que 0 es el que mas repite
ndf['X9'].value_counts()

 0.0    16450
-1.0     5685
-2.0     4348
 2.0     3157
 3.0      180
 4.0       69
 7.0       58
 5.0       35
 6.0        5
 1.0        2
 8.0        2
Name: X9, dtype: int64

In [238]:
ndf['X9'].fillna(0.0, inplace = True)

In [239]:
#El valor del vacio del ID 23993 de la columna X9 Fue reemplazado por 0
ndf[ndf['ID'] == 23993]

ID     X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15  \
23992  23993  50000  1.0  2.0  2.0  24.0  0.0  0.0  0.0  0.0  ...  41286.0   

           X16      X17     X18     X19     X20    X21    X22    X23    Y  
23992  16764.0  16960.0  2150.0  2352.0  1239.0  717.0  585.0  589.0  0.0  

[1 rows x 25 columns]

In [240]:
#Se revisa cuantos datos vacios llevamos
ndf.isna().sum()

ID      0
X1      0
X2      0
X3      0
X4      0
X5      0
X6      0
X7      0
X8      0
X9      0
X10    11
X11     9
X12     6
X13     6
X14     8
X15    10
X16    12
X17     8
X18     6
X19     7
X20     6
X21     9
X22     9
X23     5
Y       3
dtype: int64

En la columna X12 se detecta que los 6 row que contienen NaN tambien les faltan mas de 5 datos se toma la desicion de eliminarlos


In [241]:
ndf[ndf['X12'].isna()]

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...  X15  X16  \
18        19  360000  2.0  1.0  1.0  49.0  1.0 -2.0 -2.0 -2.0  ...  NaN  NaN   
29733  29734  300000  1.0  2.0  1.0  37.0 -2.0 -2.0 -1.0 -1.0  ...  NaN  NaN   
29734  29735  500000  1.0  2.0  1.0  49.0 -1.0 -1.0  2.0 -1.0  ...  NaN  NaN   
29735  29736   50000  1.0  3.0  1.0  40.0  2.0  0.0  0.0  0.0  ...  NaN  NaN   
29824  29825   40000  1.0  1.0  1.0  47.0  2.0  2.0  2.0  2.0  ...  NaN  NaN   
29825  29826   50000  1.0  2.0  1.0  41.0  0.0  0.0  0.0  0.0  ...  NaN  NaN   

       X17  X18  X19  X20  X21  X22     X23    Y  
18     NaN  0.0  0.0  0.0  0.0  0.0     0.0  0.0  
29733  NaN  NaN  NaN  NaN  NaN  NaN  3081.0  0.0  
29734  NaN  NaN  NaN  NaN  NaN  NaN   237.0  1.0  
29735  NaN  NaN  NaN  NaN  NaN  NaN   406.0  1.0  
29824  NaN  NaN  NaN  NaN  NaN  NaN     NaN  NaN  
29825  NaN  NaN  NaN  NaN  NaN  NaN     NaN  NaN  

[6 rows x 25 columns]

In [242]:
ndf.dropna(subset=["X12"], inplace= True)

In [243]:
#Se revisa cuantos datos vacios llevamos
ndf.isna().sum()

ID     0
X1     0
X2     0
X3     0
X4     0
X5     0
X6     0
X7     0
X8     0
X9     0
X10    8
X11    6
X12    0
X13    0
X14    2
X15    4
X16    6
X17    2
X18    1
X19    2
X20    1
X21    4
X22    4
X23    3
Y      1
dtype: int64

Para las columnas X10 y X11 Se decide utilizar el dato moda en este caso 0 en ambas columnas para reemplazar los valores perdidos

In [244]:
ndf[ndf['X10'].isna()]

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...       X15  \
49      50   20000  1.0  1.0  2.0  24.0  0.0  0.0  0.0  0.0  ...   19865.0   
233    234  190000  1.0  2.0  2.0  34.0  2.0  0.0  0.0  0.0  ...  142323.0   
6268  6269  120000  1.0  2.0  2.0  29.0  1.0  2.0  0.0  0.0  ...   91380.0   
6269  6270  400000  2.0  1.0  2.0  27.0  1.0 -2.0 -1.0  0.0  ...   12640.0   
6270  6271   50000  2.0  1.0  1.0  55.0  0.0  0.0  0.0  0.0  ...   19431.0   
6382  6383   80000  2.0  2.0  1.0  30.0  1.0  2.0  0.0  0.0  ...   75374.0   
6383  6384   90000  1.0  2.0  2.0  29.0  0.0  0.0  0.0  0.0  ...   69362.0   
6384  6385   20000  2.0  2.0  2.0  22.0  0.0  0.0  0.0  0.0  ...   20055.0   

           X16       X17     X18      X19      X20     X21      X22     X23  \
49     20480.0   20063.0  1318.0   1315.0    704.0   928.0    912.0  1069.0   
233   140120.0  150052.0  5000.0   5000.0  10000.0     0.0  12118.0  2769.0   
6268   93263.0   95079.0     0.0   3200.0   6800.0  3500.0   3500.0     0.0   
6269   14805.0     238.0   140.0  12527.0    253.0  2305.0    238.0   238.0   
6270   19837.0   20232.0  1775.0   1815.0    695.0   719.0    724.0   750.0   
6382   77158.0   78710.0     0.0   3299.0   2699.0  3000.0   2987.0  3100.0   
6383   70537.0   56775.0  4400.0   4111.0   2100.0  2200.0   2510.0  2000.0   
6384   19606.0   19925.0  1298.0   1909.0   1666.0  5000.0    703.0   775.0   

        Y  
49    0.0  
233   1.0  
6268  1.0  
6269  0.0  
6270  0.0  
6382  0.0  
6383  0.0  
6384  0.0  

[8 rows x 25 columns]

In [245]:
ndf['X10'].mode()

0    0.0
dtype: float64

In [246]:
ndf['X11'].mode()

0    0.0
dtype: float64

In [247]:
#Se revisa que valores contiene la columna determinando que 0 es el que mas repite
ndf['X10'].value_counts()

 0.0    16936
-1.0     5535
-2.0     4545
 2.0     2623
 3.0      178
 4.0       84
 7.0       58
 5.0       17
 6.0        4
 8.0        1
Name: X10, dtype: int64

In [248]:
#Se revisa que valores contiene la columna determinando que 0 es el que mas repite
ndf['X11'].value_counts()

 0.0    16277
-1.0     5735
-2.0     4894
 2.0     2764
 3.0      184
 4.0       49
 7.0       46
 6.0       19
 5.0       13
 8.0        2
Name: X11, dtype: int64

In [249]:
ndf['X10'].fillna(0.0, inplace = True)
ndf['X11'].fillna(0.0, inplace = True)

In [250]:
#Se revisa cuantos datos vacios llevamos
ndf.isna().sum()

ID     0
X1     0
X2     0
X3     0
X4     0
X5     0
X6     0
X7     0
X8     0
X9     0
X10    0
X11    0
X12    0
X13    0
X14    2
X15    4
X16    6
X17    2
X18    1
X19    2
X20    1
X21    4
X22    4
X23    3
Y      1
dtype: int64

En la columna X14 se reemplazo por el valor 0 debido a que es el valor moda

In [251]:
ndf[ndf['X14'].isna()]

ID     X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15  \
6232    6233  60000  2.0  2.0  2.0  29.0  2.0  2.0  2.0  0.0  ...      NaN   
17990  17991  50000  1.0  3.0  2.0  41.0  0.0  0.0  0.0  0.0  ...  50269.0   

           X16      X17     X18     X19     X20     X21     X22     X23    Y  
6232       NaN      NaN     NaN     NaN     NaN     NaN     NaN     NaN  NaN  
17990  29182.0  29096.0  4200.0  2200.0  3300.0  1200.0  1100.0  1100.0  0.0  

[2 rows x 25 columns]

In [252]:
ndf['X14'].value_counts()

0.0         2868
390.0        275
780.0         73
326.0         63
316.0         62
            ... 
12438.0        1
19580.0        1
45129.0        1
227807.0       1
49764.0        1
Name: X14, Length: 22020, dtype: int64

In [253]:
ndf['X14'].fillna(0.0, inplace = True)

In [254]:
#Se revisa cuantos datos vacios llevamos
ndf.isna().sum()

ID     0
X1     0
X2     0
X3     0
X4     0
X5     0
X6     0
X7     0
X8     0
X9     0
X10    0
X11    0
X12    0
X13    0
X14    0
X15    4
X16    6
X17    2
X18    1
X19    2
X20    1
X21    4
X22    4
X23    3
Y      1
dtype: int64

In [255]:

ndf[ndf['X15'].isna()]

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...  X15  X16  \
6232  6233   60000  2.0  2.0  2.0  29.0  2.0  2.0  2.0  0.0  ...  NaN  NaN   
6276  6277   30000  1.0  3.0  1.0  32.0  0.0  0.0  0.0  0.0  ...  NaN  NaN   
6277  6278  110000  2.0  1.0  2.0  32.0  1.0  2.0  0.0  0.0  ...  NaN  NaN   
6278  6279   20000  2.0  3.0  2.0  54.0  0.0  0.0  2.0  2.0  ...  NaN  NaN   

          X17     X18     X19     X20     X21    X22     X23    Y  
6232      NaN     NaN     NaN     NaN     NaN    NaN     NaN  NaN  
6276   1400.0  3000.0  1431.0   483.0     0.0  700.0     0.0  0.0  
6277  25453.0  1210.0  2056.0  1200.0  2300.0    0.0  1000.0  1.0  
6278   9116.0  2000.0  1000.0  1000.0     0.0  500.0   500.0  0.0  

[4 rows x 25 columns]

In [256]:
ndf['X15'].value_counts()

0.0         3194
390.0        245
780.0        101
316.0         68
326.0         62
            ... 
118839.0       1
23567.0        1
18377.0        1
24827.0        1
36535.0        1
Name: X15, Length: 21541, dtype: int64

In [257]:
ndf['X16'].value_counts()

0.0        3504
390.0       235
780.0        94
316.0        79
326.0        62
           ... 
19341.0       1
66726.0       1
80682.0       1
28508.0       1
32428.0       1
Name: X16, Length: 21000, dtype: int64

In [258]:
ndf['X15'].fillna(0.0, inplace = True)
ndf['X16'].fillna(0.0, inplace = True)

In [259]:
#Se revisa cuantos datos vacios llevamos
ndf.isna().sum()

ID     0
X1     0
X2     0
X3     0
X4     0
X5     0
X6     0
X7     0
X8     0
X9     0
X10    0
X11    0
X12    0
X13    0
X14    0
X15    0
X16    0
X17    2
X18    1
X19    2
X20    1
X21    4
X22    4
X23    3
Y      1
dtype: int64

In [260]:
ndf[ndf['X17'].isna()]

ID     X1   X2   X3   X4    X5   X6   X7   X8   X9  ...  X15  X16  \
38      39  50000  1.0  1.0  2.0  25.0  1.0 -1.0 -1.0 -2.0  ...  0.0  0.0   
6232  6233  60000  2.0  2.0  2.0  29.0  2.0  2.0  2.0  0.0  ...  0.0  0.0   

      X17    X18  X19  X20  X21  X22  X23    Y  
38    NaN  780.0  0.0  0.0  0.0  0.0  0.0  1.0  
6232  NaN    NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[2 rows x 25 columns]

In [261]:
ndf['X17'].value_counts()

0.0        4018
390.0       206
780.0        86
150.0        78
316.0        77
           ... 
16139.0       1
1033.0        1
18814.0       1
15892.0       1
15313.0       1
Name: X17, Length: 20598, dtype: int64

In [262]:
ndf['X17'].fillna(0.0, inplace = True)

In [263]:
#Se revisa cuantos datos vacios llevamos
ndf.isna().sum()


ID     0
X1     0
X2     0
X3     0
X4     0
X5     0
X6     0
X7     0
X8     0
X9     0
X10    0
X11    0
X12    0
X13    0
X14    0
X15    0
X16    0
X17    0
X18    1
X19    2
X20    1
X21    4
X22    4
X23    3
Y      1
dtype: int64

In [264]:
ndf[ndf['X18'].isna()]

ID     X1   X2   X3   X4    X5   X6   X7   X8   X9  ...  X15  X16  \
6232  6233  60000  2.0  2.0  2.0  29.0  2.0  2.0  2.0  0.0  ...  0.0  0.0   

      X17  X18  X19  X20  X21  X22  X23   Y  
6232  0.0  NaN  NaN  NaN  NaN  NaN  NaN NaN  

[1 rows x 25 columns]

Para la columna X18 se elimna el row con ID 6233 debido a contiene una gran cantidad de datos perdidos

In [265]:
ndf.dropna(subset=["X18"], inplace= True)

In [266]:
#Se revisa cuantos datos vacios llevamos
ndf.isna().sum()

ID     0
X1     0
X2     0
X3     0
X4     0
X5     0
X6     0
X7     0
X8     0
X9     0
X10    0
X11    0
X12    0
X13    0
X14    0
X15    0
X16    0
X17    0
X18    0
X19    1
X20    0
X21    3
X22    3
X23    2
Y      0
dtype: int64

In [267]:
ndf[ndf['X19'].isna()]

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...  X15     X16  \
64  65  130000  2.0  2.0  1.0  51.0 -1.0 -1.0 -2.0 -2.0  ...  0.0  2353.0   

    X17  X18  X19  X20     X21  X22  X23    Y  
64  0.0  0.0  NaN  0.0  2353.0  0.0  0.0  0.0  

[1 rows x 25 columns]

In [268]:
ndf['X19'].value_counts()

0.0        5393
2000.0     1289
3000.0      856
5000.0      717
1000.0      594
           ... 
7505.0        1
7866.0        1
21300.0       1
24104.0       1
3526.0        1
Name: X19, Length: 7899, dtype: int64

In [269]:
ndf['X19'].fillna(0.0, inplace = True)

In [270]:
#Se revisa cuantos datos vacios llevamos
ndf.isna().sum()

ID     0
X1     0
X2     0
X3     0
X4     0
X5     0
X6     0
X7     0
X8     0
X9     0
X10    0
X11    0
X12    0
X13    0
X14    0
X15    0
X16    0
X17    0
X18    0
X19    0
X20    0
X21    3
X22    3
X23    2
Y      0
dtype: int64

In [271]:
ndf['X21'].value_counts()

0.0        6403
1000.0     1394
2000.0     1214
3000.0      887
5000.0      810
           ... 
9080.0        1
73319.0       1
28740.0       1
19502.0       1
8049.0        1
Name: X21, Length: 6936, dtype: int64

In [272]:
ndf['X22'].value_counts()

0.0        6700
1000.0     1340
2000.0     1322
3000.0      947
5000.0      814
           ... 
409.0         1
9111.0        1
16496.0       1
4819.0        1
52964.0       1
Name: X22, Length: 6894, dtype: int64

In [273]:
ndf['X23'].value_counts()

0.0        7169
1000.0     1298
2000.0     1295
3000.0      914
5000.0      808
           ... 
7329.0        1
6862.0        1
6525.0        1
11894.0       1
16080.0       1
Name: X23, Length: 6938, dtype: int64

Se revisaron las columnas X21,X22,X23 y solo presentaban un dato faltante se decidio utilizar el valor moda de cada columna para rellenar los datos faltantes

In [274]:
ndf['X21'].fillna(0.0, inplace = True)
ndf['X22'].fillna(0.0, inplace = True)
ndf['X23'].fillna(0.0, inplace = True)

In [275]:
#Ya ningun dato contiene elementos vacios
ndf.isna().sum()

ID     0
X1     0
X2     0
X3     0
X4     0
X5     0
X6     0
X7     0
X8     0
X9     0
X10    0
X11    0
X12    0
X13    0
X14    0
X15    0
X16    0
X17    0
X18    0
X19    0
X20    0
X21    0
X22    0
X23    0
Y      0
dtype: int64

In [276]:
ndf.describe()

ID              X1            X2            X3            X4  \
count  29988.000000    29988.000000  29988.000000  29988.000000  29988.000000   
mean   14997.515906   167497.655062      1.603908      1.853075      1.551954   
std     8658.125501   129735.297746      0.489092      0.790375      0.521971   
min        1.000000    10000.000000      1.000000      0.000000      0.000000   
25%     7500.750000    50000.000000      1.000000      1.000000      1.000000   
50%    14997.500000   140000.000000      2.000000      2.000000      2.000000   
75%    22494.250000   240000.000000      2.000000      2.000000      2.000000   
max    30000.000000  1000000.000000      2.000000      6.000000      3.000000   

                 X5            X6            X7            X8            X9  \
count  29988.000000  29988.000000  29988.000000  29988.000000  29988.000000   
mean      35.483308     -0.016740     -0.133687     -0.166533     -0.220755   
std        9.217860      1.123714      1.197159      1.195895      1.169079   
min       21.000000     -2.000000     -2.000000     -2.000000     -2.000000   
25%       28.000000     -1.000000     -1.000000     -1.000000     -1.000000   
50%       34.000000      0.000000      0.000000      0.000000      0.000000   
75%       41.000000      0.000000      0.000000      0.000000      0.000000   
max       79.000000      8.000000      8.000000      8.000000      8.000000   

       ...            X15            X16           X17            X18  \
count  ...   29988.000000   29988.000000   29988.00000   29988.000000   
mean   ...   43271.323029   40317.770542   38881.71762    5663.592704   
std    ...   64343.737119   60807.144157   59563.42570   16565.173454   
min    ... -170000.000000  -81334.000000 -339603.00000       0.000000   
25%    ...    2324.500000    1761.000000    1256.00000    1000.000000   
50%    ...   19051.000000   18104.500000   17075.50000    2100.000000   
75%    ...   54558.500000   50207.500000   49215.25000    5006.000000   
max    ...  891586.000000  927171.000000  961664.00000  873552.000000   

                X19            X20            X21            X22  \
count  2.998800e+04   29988.000000   29988.000000   29988.000000   
mean   5.922910e+03    5226.042617    4827.200747    4799.768341   
std    2.304529e+04   17609.531285   15669.025313   15280.888462   
min    0.000000e+00       0.000000       0.000000       0.000000   
25%    8.357500e+02     390.000000     296.000000     251.000000   
50%    2.009000e+03    1800.000000    1500.000000    1500.000000   
75%    5.000000e+03    4505.500000    4014.000000    4031.500000   
max    1.684259e+06  896040.000000  621000.000000  426529.000000   

                 X23             Y  
count   29988.000000  29988.000000  
mean     5217.108910      0.221155  
std     17780.813339      0.415032  
min         0.000000      0.000000  
25%       116.750000      0.000000  
50%      1500.000000      0.000000  
75%      4000.000000      0.000000  
max    528666.000000      1.000000  

[8 rows x 25 columns]